File: Atcoder_2024/0420/d/main.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;



def find(int i , vector<vector<int>> cluster){
    while( i != cluster[i]){

    }
}


int main() {   
    int n, m;
    cin >> n >> m;

    vector<vector<int>> a(n);

    int x, y;
    for (int i = 0; i < m; i++) {
        cin >> x >> y;
        x--; 
        y--; 
        if (x>y){
            a[x].push_back(y);
        }
        else{
            a[y].push_back(x);
        }
    }

    int ans = 0;



    for (int i = 0; i < n; i++) {
        while (i != a[i]) {
            i = a[i];
        }
    }

    return 0;
}

File: Atcoder_2024/0420/d/main.py


def find(i,cluster):
    while i != cluster[i]:
        i = cluster[i]
    return i

n,m = map(int,input().split())
a = [ [] for _ in range(n)]

for i in range(m):
    x,y = map(int,input().split())
    x -=1
    y -=1
    if x < y:
        a[y].append(x)
    else: a[x].append(y)
    
ans = 0
for i in range(n):
    x = find(i,a[i])
    for j in a[i]:
        y = find(j,a[j])
        if x != y:
            ans += 1
print(ans)
        
    
